src/backend/distributed/metadata/node_metadata.c

```diff
@@ -1152,100 +934,100 @@ ActivateNodeList(List *nodeList)
+	/*
+	 * we need to unset metadatasynced flag to false at coordinator in separate
+	 * transaction only at nontransactional sync mode and if we do not collect
+	 * commands.
+	 *
+	 * We make sure we set the flag to false at the start of nontransactional
+	 * metadata sync to mark those nodes are not synced in case of a failure in
+	 * the middle of the sync.
+	 */
+	if (context->transactionMode == METADATA_SYNC_NON_TRANSACTIONAL &&
+		!MetadataSyncCollectsCommands(context))
 	{
-		/*
-		 * First, locally mark the node is active, if everything goes well,
-		 * we are going to sync this information to all the metadata nodes.
-		 */
-		WorkerNode *workerNode =
-			FindWorkerNodeAnyCluster(node->workerName, node->workerPort);
-		if (workerNode == NULL)
-		{
-			ereport(ERROR, (errmsg("node at \"%s:%u\" does not exist", node->workerName,
-								   node->workerPort)));
-		}
+		MarkNodesNotSyncedInSeparateTransaction(context);
+	}
 
-		/* both nodes should be the same */
-		Assert(workerNode->nodeId == node->nodeId);
+	/*
+	 * Take an exclusive lock on pg_dist_node to serialize pg_dist_node
+	 * changes. We should not acquire the lock before deactivating
+	 * metadata nodes as it causes deadlock.
+	 */
+	LockRelationOid(DistNodeRelationId(), ExclusiveLock);
 
-		/*
-		 * Delete existing reference and replicated table placements on the
-		 * given groupId if the group has been disabled earlier (e.g., isActive
-		 * set to false).
-		 *
-		 * Sync the metadata changes to all existing metadata nodes irrespective
-		 * of the current nodes' metadata sync state. We expect all nodes up
-		 * and running when another node is activated.
-		 */
-		if (!workerNode->isActive && NodeIsPrimary(workerNode))
-		{
-			bool localOnly = false;
-			DeleteAllReplicatedTablePlacementsFromNodeGroup(workerNode->groupId,
-															localOnly);
-		}
+	/*
+	 * Error if there is concurrent change to node table before acquiring
+	 * the lock
+	 */
+	ErrorIfAnyNodeNotExist(context->activatedWorkerNodeList);
 
-		workerNode =
-			SetWorkerColumnLocalOnly(workerNode, Anum_pg_dist_node_isactive,
-									 BoolGetDatum(true));
+	/*
+	 * Delete existing reference and replicated table placements on the
+	 * given groupId if the group has been disabled earlier (e.g., isActive
+	 * set to false).
+	 * todo: use metada context connections
+	 */
+	SendDeletionCommandsForReplicatedTablePlacements(context);
 
-		/* TODO: Once all tests will be enabled for MX, we can remove sync by default check */
-		bool syncMetadata = EnableMetadataSync && NodeIsPrimary(workerNode);
-		if (syncMetadata)
+	/*
+	 * We are going to sync the metadata anyway in this transaction, so do
+	 * not fail just because the current metadata is not synced.
+	 * (ErrorIfAnyMetadataNodeOutOfSync causes failure)
+	 *
+	 * Set also isactive to true locally so that we can find active workers.
+	 * The changes will rollback in case of failure anyway.
+	 */
+	if (!MetadataSyncCollectsCommands(context))
+	{
```

Comment:
would be nice to move this logic to a function: `SetNodeMetadataLocally(context)` or such